include/fbgemm/OutputProcessing-inl.h

```diff
@@ -44,11 +44,270 @@ inline int DoSpmdmOnInpBuffer<outT, inT, nextOPType>::f(
   return nextop_.template f<instSet>(out, inp, block, ld_out, ld_in);
 }
 
+template <bool FUSE_RELU, typename outT, typename inT, typename nextOPType>
+template <bool A_SYMMETRIC, bool B_SYMMETRIC, bool HAS_BIAS>
+void ReQuantizeOutput<FUSE_RELU, outT, inT, nextOPType>::f_(
+    outT* out,
+    const inT* inp,
+    const block_type_t& block,
+    int ld_out,
+    int ld_in) const {
+  // Adoption of implementation at QNNPACK/src/requantization/fp32-sse2.c
+  // using AVX2 instructions
+  __m256 multiplier_v = _mm256_set1_ps(C_multiplier_);
+
+  __m256i min_v = _mm256_set1_epi8(std::numeric_limits<uint8_t>::min());
+  __m256i max_v = _mm256_set1_epi8(std::numeric_limits<uint8_t>::max());
+
+  assert(
+      (A_SYMMETRIC == (Aq_zero_point_ == 0)) &&
+      "A_SYMMETRIC == true if and only if Aq_zero_point == 0");
+  assert(
+      (B_SYMMETRIC == (Bq_zero_point_ == 0 || q_row_offsets_ == nullptr)) &&
+      "B_SYMMETRIC == true if and only if Bq_zero_point == 0 "
+      "or q_row_offsets_ == nullptr");
+  assert(
+      (HAS_BIAS == (bias_ != nullptr)) &&
+      "HAS_BIAS == true if and only if bias != nullptr");
+
+  __m256i A_zero_point_v = _mm256_set1_epi32(Aq_zero_point_);
+  __m256i C_zero_point_epi16_v = _mm256_set1_epi16(C_zero_point_);
+  __m256i C_zero_point_epi8_v = _mm256_set1_epi8(C_zero_point_);
+
+  __m256i permute_mask_v =
+      _mm256_set_epi32(0x07, 0x03, 0x06, 0x02, 0x05, 0x01, 0x04, 0x00);
+
+  constexpr int VLEN = 8;
+  for (int i = block.row_start; i < block.row_start + block.row_size; ++i) {
+    std::int32_t row_offset = (q_row_offsets_ && !B_SYMMETRIC)
+        ? q_row_offsets_[i - block.row_start] * Bq_zero_point_
+        : 0;
+    __m256i row_offset_v = _mm256_set1_epi32(row_offset);
+    int j = block.col_start;
+    for (; j < block.col_start + (block.col_size / (VLEN * 4) * (VLEN * 4));
+         j += (VLEN * 4)) {
+      __m256i x_v = _mm256_loadu_si256(reinterpret_cast<const __m256i*>(
+          inp + (i - block.row_start) * ld_in + (j - block.col_start)));
+      __m256i y_v = _mm256_loadu_si256(reinterpret_cast<const __m256i*>(
+          inp + (i - block.row_start) * ld_in + (j - block.col_start) +
+          1 * VLEN));
+      __m256i z_v = _mm256_loadu_si256(reinterpret_cast<const __m256i*>(
+          inp + (i - block.row_start) * ld_in + (j - block.col_start) +
+          2 * VLEN));
+      __m256i w_v = _mm256_loadu_si256(reinterpret_cast<const __m256i*>(
+          inp + (i - block.row_start) * ld_in + (j - block.col_start) +
+          3 * VLEN));
+
+      if (!A_SYMMETRIC) {
```

Comment:
You can always force `A_zero_point` to be zero by fusing the effect of `K * A_zero_point * B_zero_point - sum(B[i][.]) * A_zero_point` into bias. QNNPACK now does this optimization.